### ISSUE TITLE
Removing the Application contracts

### ISSUE THREAD

**Issue by driesvints** [laravel/ideas#1772]
Couldn't find an issue for this so I thought I'd start a new one here.

In a sense I'd like to gather some opinions on the point of the `Illuminate\Contracts\Foundation\Application` and `Illuminate\Contracts\Console\Application` contracts. To me personally, they don't serve any purpose other than "swap out the Laravel framework with another framework" or "swap out Artisan with another CLI tool". I think this goes against exactly why you'd use a dedicated framework in the first place rather than opting to use different Illuminate components to build up your toolchain.

Especially the `Illuminate\Contracts\Foundation\Application` contract has become really bloated and basically requires you to re-implement the entire framework yourself. The fact that it extends the `Container` contract is just a pure violation of it core purpose. Interfaces aren't meant to be extended like this if they don't share a common purpose. In this case the `Application` contract defines what Laravel is and how it can be used and the `Container` contract just defines how a container should be implemented. The fact that the application works as a container proxy/implementation is besides the point. 

What I'd opt for it to remove both the `Illuminate\Contracts\Foundation\Application` and `Illuminate\Contracts\Console\Application` contracts and use more sensible dependencies where possible. Trying to find out where a pure `Container` would be a better dependency than the whole `Application` would go a long way to de-couple more components.

Another thing we could do is to see how we could split up some of the things from the `Illuminate\Contracts\Foundation\Application` into separate smaller contracts. For example, the `isDownForMaintenance` is used in several places and could easily be put into its own dedicated (and smaller) contract:

```php
interface MaintenanceChecker
{
    public function isDownForMaintenance(): bool;
}
```

This could then be implemented by the `Illuminate\Foundation\Application` class, added to the core aliases and auto-resolved from the container. There's much more we can move to separate contracts like the environment checks, the service provider methods, the application path methods, etc.

So feel free to leave some thoughts here. Especially if you have use cases for the `Application` contracts.

**Comment by driesvints** [comment]
Follow up: it actually also occurs to me that `Illuminate\Foundation\Application` is just a bad name overall. A better name would probably be `Illuminate\Foundation\Laravel`.

**Comment by driesvints** [comment]
@pavinthan if you can react with a thumbs down then can you please also share why you'd want to keep them?

**Comment by stefanbauer** [comment]
I agree.

> Trying to find out where a pure Container would be a better dependency than the whole Application would go a long way to de-couple more components.

I would, as you mentioned, go with the smallest possible type of injection. Even the container might often be way too much. With your given example by the MaintenanceChecker Interface, I think that's exactly the direction I would go. 👍

**Comment by patrickbrouwers** [comment]
I think in the past I've used the contract interface to have an app specific implementation of Application contract. I believe it was to overrule the paths methods in very early Laravel 5 versions.

Having smaller contracts sound like the better solution. (ISP) Could at least be a good step before while deprecation the Application contract.

**Comment by driesvints** [comment]
@patrickbrouwers would you still need that override in the current state of the framework?

**Comment by patrickbrouwers** [comment]
@driesvints I see a lot of setters were added over time to deal with that. 
I think we used it to change `public` to `public_html` because that was the default our old vps servers had. But that's not really an issue for us anymore.

I've checked the most recent code bases and we don't really do anything with the Application contract anymore. Only see usages from: laravel itself, debugbar, intervention, telescope, larastan and orchestra testbench. But it doesn't seem like any of those depend on swapping out the application instance.

**Comment by peterfox** [comment]
I agree that breaking down the contracts makes a lot of sense.

**Comment by JosephSilber** [comment]
@patrickbrouwers did you actually have a full-on app specific re-implementation of the contract, or you merely extended the `Application` class?

**Comment by patrickbrouwers** [comment]
It was just an extension.

**Comment by JosephSilber** [comment]
@patrickbrouwers I figured (I've extended it myself before).

Removing the contract won't prevent you from extending it 👍

**Comment by patrickbrouwers** [comment]
@JosephSilber that's true. You can still extend it and re-bind the application instance. It won't really change anything in that regard.

**Comment by franzliedke** [comment]
I agree the contract is mostly useless in its current form - especially because I often saw inconsistent usages of the contract vs. the implementation (e.g. docblocks claiming the contract was used, but the code used methods from the implementation). That may have changed, but if so, probably just by adding even more methods to the contract.

Removing it is one solution to the "problem", but doesn't solve the root problem, which to me is too many responsibilities in the class *and* contract.

Interface segregation, which you've also suggested, seems like the most sensible route to me. :+1:

And to avoid the "meh, this is just theoretical nitpicking": it would be very useful for me in Flarum, which uses many components and wants to be easy to integrate with the framework, without using the full-stack thing. The Application class / contract has been one of the biggest pain-points.

**Comment by olivernybroe** [comment]
As a package developer I think splitting it to multiple interfaces is a nice way to go. 
Might even help a lot with the inconsistencies in Laravel and Lumen and make it easy for other frameworks to add support for my package as they just need to implement a few methods of a interface instead of the whole Application interface.

**Comment by crynobone** [comment]
A similar discussion was made previously under laravel/framework#26477

**Comment by franzliedke** [comment]
For the record: custom implementations of that contract do exist. laravel/framework#26477 mentions one, I have also built one before.
